### DISCUSSION TITLE
Will Pydantic support LiteralString?

### DISCUSSION THREAD

**Discussion by andrew222651** [pydantic/pydantic#4574] [Question]
Currently, this
```python
from pydantic import BaseModel, Field
from typing_extensions import LiteralString

class Foo(BaseModel):
    s: LiteralString
```
generates an error:
```python
  ...
  File "pydantic/validators.py", line 718, in find_validators
RuntimeError: error checking inheritance of typing_extensions.LiteralString (type: LiteralString)
```

See <https://peps.python.org/pep-0675/> for info on `LiteralString`. Would there be any issues due to `parse_obj`, etc? My own use case is using `LiteralString` for a field in an ABC and using `Literal` types in the concrete subclasses. A hack which (I think?) works is `s = Field()  # type: LiteralString`.

Sept 22, 2023 addendum: The usage described here violates `reportIncompatibleVariableOverride` in Pyright.

**Comment by tibbe** [comment]
I have exactly the same use-case and issue.

**Reply by dmontagu** [reply]
Without digging into the details of how it works, I suspect we would need to treat it at runtime as equivalent to `str`, because I don't believe there is any way at runtime to determine if a string was a bound reference or a literal. (Happy to be corrected if this is wrong.)

That said, if we do just treat it a string from the perspective of runtime parsing/etc., you'd presumably still get the right type-checking from mypy (with the obvious exception of parsing methods).

I think the parsing-related functionality is the biggest unknown for what the right behavior is, considering I would imagine it would be functionally impossible to ensure that the parsing method was called on a string literal. I could imagine a world where we set it up so that parsing *always* fails for this type, but I think it may be difficult to achieve cleanly in a way that doesn't impact performance for those who do not use this type.

I would be open to opinions about how this should be done, but just want to make the point that, to the extent that these things end up in conflict, keeping the rest of the library maintainable will probably take priority over perfectly handling the semantics of LiteralString, even if it is theoretically possible to do so.

**Reply by tibbe** [reply]
My intuition was also to make parsing of this type always fail, expecting that it will always be overridden in subclasses.

**Reply by tibbe** [reply]
Ran into this today when trying to enable strict type checking using Pylance in VSCode.

We have discriminate unions that look like this:

```py
class BaseItem(BaseModel):
  Type: str

class A(BaseItem):
  Type: Literal['A']

class B(BaseItem):
  Type: Literal['B']
```

Pylance doesn't like this and complains:

```
"Type" overrides symbol of same name in class "BaseItem"
  Variable is mutable so its type is invariant
    Override type "Literal['A']" is not the same as base type "str"
```

Ideally we'd like to change the type of `Type` in `BaseItem` to be `LiteralString` but pydantic doesn't like that.

**Reply by andrew222651** [reply]
@tibbe I believe even if you had `LiteralString` instead of `str` there, Pydantic would still report that error

ETA: Pyright, not Pydantic

**Reply by tibbe** [reply]
> @tibbe I believe even if you had `LiteralString` instead of `str` there, Pydantic would still report that error

Yes, that's the issue. What I did here was to provide further reason why this issue is worth fixing.

**Reply by andrew222651** [reply]
Sorry >< I meant Pyright

**Reply by tibbe** [reply]
> Sorry >< I meant Pyright

Ah. Will have to experiment.

**Reply by andrew222651** [reply]
Pyright is OK with the following. Pydantic will complain but it will work.
```python
_T1 = TypeVar("_T1", bound=LiteralString)

class Foo(BaseModel, abc.ABC, Generic[_T1]):
    s: _T1
    model_config = ConfigDict(arbitrary_types_allowed=True, frozen=True)

class Bar(Foo[Literal["baz"]]):
    s: Literal["baz"]
```

**Reply by tibbe** [reply]
Is the `frozen` required here?

**Reply by tibbe** [reply]
Btw the reason I want to do this to begin with is that `BaseItem` (your `Foo`) has methods implemented in terms of the `Type` attribute (your `s`) and thus I want the attribute defined there.

My `BaseItem` is a base class for an ORM on top of DynamoDB.

**Reply by andrew222651** [reply]
No, I don't think `frozen` is relevant. (If Pyright understood what it means, we wouldn't need generics but of course it doesn't.)

**Reply by mixam24** [reply]
I slightly modified the code snippet of @amacfie  to make happy both Pyright and Pydantic and get rid off `arbitrary_types_allowed=True` property:

```
from pydantic import BaseModel, GetPydanticSchema

PydanticFriendlyStringLiteral = Annotated[LiteralString, GetPydanticSchema(lambda _s, h: h(str))]

_T1 = TypeVar("_T1", bound=PydanticFriendlyStringLiteral)

class Foo(BaseModel, Generic[_T1]):
    s: _T1


class Bar(Foo[Literal["baz"]]):
    s: Literal["baz"]
```
